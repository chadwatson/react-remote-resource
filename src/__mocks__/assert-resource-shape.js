/**
 * Validates that the resource passed in has the following structure:
 *
 * {
 *   getState: `function`,
 *   setState: `function`,
 *   refresh: `function`,
 *   useState: `function`,
 *   subscribe: `function`
 * }
 *
 * @param {object} resource
 */
export const assertResourceShape = resource => {
  expect(typeof resource.id).toBe("string");
  expect(typeof resource.getState).toBe("function");
  expect(typeof resource.setState).toBe("function");
  expect(typeof resource.refresh).toBe("function");
  expect(typeof resource.useState).toBe("function");
  expect(typeof resource.subscribe).toBe("function");
};
