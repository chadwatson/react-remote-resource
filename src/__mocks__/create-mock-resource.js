import createResource from "../create-resource";

export const createMockResource = (
  getter,
  setter,
  loader,
  entriesExpireAfter
) => {
  const spies = {
    getter: jest.fn().mockImplementation(getter),
    setter: jest.fn().mockImplementation(setter),
    loader: jest.fn().mockImplementation(loader)
  };

  return [
    createResource(
      spies.getter,
      spies.setter,
      spies.loader,
      entriesExpireAfter
    ),
    spies
  ];
};
