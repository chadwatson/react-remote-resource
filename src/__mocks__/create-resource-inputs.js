export const createResourceInputs = () => ({
  getter: jest.fn().mockImplementation(state => state),
  setter: jest.fn().mockImplementation((_, __, newState) => newState),
  entryPredicate: jest.fn().mockImplementation(state => !!state),
  loader: jest.fn().mockImplementation(value => Promise.resolve(value))
});
