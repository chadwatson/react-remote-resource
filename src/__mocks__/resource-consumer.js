import { useEffect } from "react";
import useEntry from "../use-entry";

export const ResourceConsumer = ({ resource, startValue, endValue }) => {
  const [value, setValue] = useEntry(resource, [startValue]);

  useEffect(() => {
    const timeout = endValue && setTimeout(() => setValue(endValue), 1000);
    return () => timeout && clearTimeout(timeout);
  }, [endValue]);

  return value;
};
