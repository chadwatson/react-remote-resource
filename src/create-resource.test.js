import createResource from "./create-resource";

import { createResourceInputs as createSpies } from "./__mocks__/create-resource-inputs";

// ---------------------------
// Tests
// ---------------------------

describe("createResource", () => {
  it("is a function", () => {
    expect(typeof createResource).toBe("function");
  });

  it("is returns a function when passed a getter", () => {
    const { getter } = createSpies();
    expect(typeof createResource(getter)).toBe("function");
  });

  it("is returns a function when passed a getter and setter", () => {
    const { getter, setter } = createSpies();
    expect(typeof createResource(getter, setter)).toBe("function");
  });

  it("is returns a function when passed a getter, setter, and entryPredicate", () => {
    const { getter, setter, entryPredicate } = createSpies();
    expect(typeof createResource(getter, setter, entryPredicate)).toBe(
      "function"
    );
  });

  it("is returns a resource when passed a getter, setter, entryPredicate, and loader", () => {
    const { getter, setter, entryPredicate, loader } = createSpies();
    const resource = createResource(getter, setter, entryPredicate, loader);
    expect(
      Object.keys(resource).map(key => [key, typeof resource[key]])
    ).toMatchSnapshot();
  });

  it("is returns a resource when passed a getter, setter, entryPredicate, loader, and getInitialState", () => {
    const { getter, setter, entryPredicate, loader } = createSpies();
    const resource = createResource(getter, setter, entryPredicate, loader);
    expect(
      Object.keys(resource).map(key => [key, typeof resource[key]])
    ).toMatchSnapshot();
  });
});

/**

const storageKey = "SYSTEMS";

// const createResource = curryN(4, (..., getInitialState = Promise.reject) => {
//   // ...
// })

const systemsResource = createResource(
  (currentState = {}, [authToken, customerId]) => currentState[customerId],
  (currentState = {}, [authToken, customerId], systems) => ({
    ...currentState,
    [customerId]: systems
  }),
  (currentState = {}, [authToken, customerId]) => typeof currentState[customerId] !== "undefined",
  (authToken, customerId) => fetch(`/v2/customers/${customerId}/control_systems?auth_token=${authToken}`)
);

systemsResource.subscribe(() => {
  localforage.setItem(storageKey, serialize(resource.getState()));
});

const Systems = () => {
  const [systems, setSystems] = useEntry(systemsResource, [authToken, customerId]);
}
 */
