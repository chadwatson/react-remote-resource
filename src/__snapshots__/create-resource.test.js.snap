// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createResource is returns a resource when passed a getter, setter, entryPredicate, and loader 1`] = `
Array [
  Array [
    "id",
    "string",
  ],
  Array [
    "getState",
    "function",
  ],
  Array [
    "setState",
    "function",
  ],
  Array [
    "refresh",
    "function",
  ],
  Array [
    "useEntry",
    "function",
  ],
  Array [
    "subscribe",
    "function",
  ],
]
`;

exports[`createResource is returns a resource when passed a getter, setter, entryPredicate, loader, and getInitialState 1`] = `
Array [
  Array [
    "id",
    "string",
  ],
  Array [
    "getState",
    "function",
  ],
  Array [
    "setState",
    "function",
  ],
  Array [
    "refresh",
    "function",
  ],
  Array [
    "useEntry",
    "function",
  ],
  Array [
    "subscribe",
    "function",
  ],
]
`;
